hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/offset/OffsetCommiter.java

```diff
@@ -0,0 +1,138 @@
+ * * if there is no min offset, commit max (all other offsets were committed)
+ * <p>
+ * This algorithm is very simple, memory efficient, can be performed in single thread and introduces no locks.
+ */
+public class OffsetCommiter implements Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger(OffsetCommiter.class);
+
+    private final ScheduledExecutorService scheduledExecutor = Executors.newSingleThreadScheduledExecutor();
+
+    private final int offsetCommitPeriodSeconds;
+
+    private final OffsetQueue offsetQueue;
+
+    private final List<MessageCommitter> messageCommitters;
+
+    private final Set<SubscriptionPartitionOffset> inflightOffsets = new HashSet<>();
+
+    private final MpscArrayQueue<SubscriptionName> subscriptionsToCleanup = new MpscArrayQueue<>(1000);
+
+    public OffsetCommiter(
+            OffsetQueue offsetQueue,
+            List<MessageCommitter> messageCommitters,
+            int offsetCommitPeriodSeconds
+    ) {
+        this.offsetQueue = offsetQueue;
+        this.messageCommitters = messageCommitters;
+        this.offsetCommitPeriodSeconds = offsetCommitPeriodSeconds;
+    }
+
+    @Override
+    public void run() {
+        offsetQueue.drainInflightOffsets(inflightOffsets::add);
+
+        Map<SubscriptionPartition, Long> maxInflightOffsets = calculateInflightOfsets(Math::max);
+
+        offsetQueue.drainCommittedOffsets(inflightOffsets::remove);
+
+        Map<SubscriptionPartition, Long> minInflightOffsets = calculateInflightOfsets(Math::min);
+
+        Set<SubscriptionPartitionOffset> offsetsToCommit = new HashSet<>();
+        maxInflightOffsets.forEach((k, v) -> {
+            if (minInflightOffsets.containsKey(k)) {
+                offsetsToCommit.add(new SubscriptionPartitionOffset(k, minInflightOffsets.get(k) - 1));
+            } else {
+                offsetsToCommit.add(new SubscriptionPartitionOffset(k, v));
+            }
+        });
+
+        for (SubscriptionPartitionOffset offset : offsetsToCommit) {
+            commit(offset);
+        }
+
+        cleanupUnusedSubscriptions();
+    }
+
+    private Map<SubscriptionPartition, Long> calculateInflightOfsets(BiFunction<Long, Long, Long> comparer) {
+        Map<SubscriptionPartition, Long> calculatedOnflightOffsets = new HashMap<>();
+        inflightOffsets.forEach(p -> calculatedOnflightOffsets.compute(
+                p.getSubscriptionPartition(),
+                (k, v) -> v == null ? p.getOffset() : comparer.apply(v, p.getOffset())
+        ));
+        return calculatedOnflightOffsets;
+    }
+
+    private void commit(SubscriptionPartitionOffset offset) {
+        for (MessageCommitter committer : messageCommitters) {
+            try {
+                committer.commitOffset(offset);
+            } catch (Exception e) {
+                logger.error("Failed to commit offset {} using {} committer", offset, committer.getClass().getSimpleName(), e);
```

Comment:
If a commit fails, we loose the information about the particular offset and we won't retry in the future. 
We should retry to commit the latest SubscriptionPartition offset in each algorithm run.